Add DevSecOps in Action: Kubernetes Secrets blog post (#45640)
kubernetes/website: content/en/blog/_posts/2024-09-17-devsecops-in-action-k8s-secrets/index.md

---
layout: blog
title: "DevSecOps in Action: Kubernetes Secrets"
slug: devsecops-in-action-k8s-secrets
date: 2024-09-17
author: >
  Evgenii Frikin
---

The Secret is one of the basic resources that exists in the Kubernetes API. This
kind of resource is used for different purposes and despite the name Secret,
sensitive data is insecure to store in it. Recently considerable attention has
been paid to
[security improvement](/docs/tasks/administer-cluster/encrypt-data/)

> for kind of Secret resource, but Kubernetes still doesn't include appropriate

Reviewer: Did you mean:```suggestion
for objects such as Secrets, but Kubernetes still doesn't include appropriate
```

Author: Applied

out-of-the-box because Kubernetes philosophy carries simplicity in order not to

> compromise it engineers have to make sure Kubernetes Secret are saved

Reviewer: I don't understand what you mean here. Can you explain?

Author: I wanted to say, that Kubernetes shouldn't include some solution for safe Secrets. No problem. I removed this part sentence.

According to
[Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices/)
there are several options on how to improve security of this kind of Secret:
encrypt secrets in the etcd, restrict the access to them via RBAC, configure
access to external Secret, etc. However, this partially solves the security issue
because they are cluster-side solutions and they help to improve security of the
cluster overall.

In the modern world, unilateral solutions aren't effective enough and can hide
security issues known as
[Security through obscurity](https://en.m.wikipedia.org/wiki/Security_through_obscurity).
To avoid it, it's necessary to integrate

> [DevSecOps methodology](https://www.redhat.com/en/topics/devops/what-is-devsecops)

Reviewer: Hi. I think you need to declare an interest here? (have I got that right?)

Author: Right. I removed hyperlink

which implies integrated security as a shared responsibility throughout the
entire IT lifecycle. In other words, security solutions must start with the
development process and apply to all other levels correspondingly.

> 

Reviewer: I think it's worth highlighting that DevSecOps is an opinion rather than a generally accepted paradigm. _You_ are welcome to buy into it; watch out for implying that Kubernetes as a project also has.

Author: Fixed

it's recommended in The
[Twelve Factors](https://12factor.net/config).
These are general recommendations which don't impact security issues that's why
their application should make sense, but passing config via environment variables
with sensitive data must be kept to a minimum.

Injection of Secret via file and in-memory has pros and cons and I would like to
show which security issues exist when using Injection of Secret via file and
suggest solutions. In fact, there will be a general solution and specific
implementation will be different in terms of specific management and delivery
secrets tools. To keep the examples simple I will use `spec.initContainers` to
simulate an injection agent for delivering secrets to Pod.

## Baseline application

Imagine there is a simple application which processes http requests. On request
to `root` location client receives environment variable value
`DEMO_SECRET__PASSWD` which is defined via
[`spec.containers.0.envFrom.secretRef.name`](kustomization/patches/baseline.yaml)
in Pod resource. On request `/readiness` receives status application
(ready or not ready) and if the environment variable isn't defined Pod doesn't
change status to `READY`.

> [`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/14624762f1546e2bfd43e65cee26186c514fa351/cmd/main.go#L10-L53)

Reviewer: We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Author: Fixed

```go
const (
	envVar = "DEMO_SECRET__PASSWD"
)

...

func httpHandle(w http.ResponseWriter, r *http.Request) {

	var isExist, ok bool
	var envVal string

	if envVal, ok = os.LookupEnv(envVar); ok {
		isExist = true
	}

	switch r.URL.Path {
	case "/":
		if isExist {
			w.WriteHeader(http.StatusOK)
			fmt.Fprintln(w, envVal)
		} else {
			w.WriteHeader(http.StatusNotFound)
			fmt.Fprintln(w, "Secret Not Found")
		}
	case "/readiness":
		if isExist {
			w.WriteHeader(http.StatusOK)
			fmt.Fprintln(w, "OK")
		} else {
			w.WriteHeader(http.StatusNotFound)
			fmt.Fprintln(w, "Environment variable wasn't defined")
		}
	default:
		w.WriteHeader(http.StatusNotFound)
		fmt.Fprintln(w, "404 Not Found")
	}
}
```

## Stage-01: Reject environment vars for storing sensitive data

According to OWASP recommendation, it's necessary to use file instead of env.
That's why it's necessary to make changes to implementation readiness probes
and deployment manifests.

> [`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/96afd59f8a391973a8ed5a9b410915b9e99d4bf4/cmd/main.go#L11-L67)

Reviewer: We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Author: Fixed

can have specific limitations. In order to reduce the amount of limitations it's
necessary to implement an approach based on support of both
functions (variables) but without additional environment variables.
This approach allows not to implement additional variable but to use the current
variable for two cases:

* legacy: using env variable for storing sensitive data
* new: using env variable for storing path to file with sensitive data and
override current value to value from the file

To distinguish sensitive data values from path to file, by using prefixes,
this will allow you to use the same variable and follow OWASP recommendations.

First let's define variables and constants:

[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L12-L20)
```go
const (
	envVar = "DEMO_SECRET__PASSWD"
	filePrefix = "file://"
)

var (
	envVal, filePath string
	isExistFile, isExistEnv, changed bool
)
```

Next, it's necessary to change the `fileExist` function. It's supposed to return
a path to the file without a prefix if the file exists.

> [`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L31-L43)

Reviewer: We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Author: Fixed

```go
func fileExist(path string) (string, bool) {
	if strings.HasPrefix(path, filePrefix) {
		path = strings.TrimPrefix(path, filePrefix)
		info, err := os.Stat(path)
		if err == nil {
			return path, !info.IsDir()
		}
		if errors.Is(err, os.ErrNotExist) {
			return "", false
		}
	}
	return "", false
}
```

Magic happens in the `httpHandle` function. Environment variable value is passed
into function, if it's a file it will be read and current environment variable
value will be overridden by a new value that was read from the file.

> [`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L45-L59)

Reviewer: We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Author: Fixed

```go
func httpHandle(w http.ResponseWriter, r *http.Request) {
	if !changed {
		if envVal, isExistEnv = os.LookupEnv(envVar); isExistEnv {
			if filePath, isExistFile = fileExist(envVal); isExistFile {
				body, err := os.ReadFile(filePath)
				if err != nil {
					log.Fatalf("unable to read file: %v", err)
				}
				os.Setenv(envVar, string(body))
				envVal = os.Getenv(envVar)
				changed = true
			}
		}
	}
}
```

Finally, if define `DEMO_SECRET__PASSWD=file:///secrets/secret.file`, deploy
application and make request to `root` location, response will contain file
content which is defined in the environment variable:

```bash
curl -v localhost:8080
```

The output is similar to this:

```text
...
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.5.0
> Accept: */*
>
< HTTP/1.1 200 OK
...
PVHBD5I21osTOX9i
```

Let's check process environment:

```bash

> # strings /proc/1/environ

Reviewer: If marking this as shell, don't include the prompt.

Author: Fixed

```

The output is similar to this:

```text
...
DEMO_SECRET__PASSWD=file:///secrets/secret.file
...
```

Excellent! Environment variable still contains the value which was defined
during the deployment.

Let's check the secret file:

```bash
# cat /secrets/secret.file

> #

Reviewer: If marking this as shell, don't include the prompt.

Author: Fixed

```

The file is empty! After readiness probe passed, the file was truncated.

If define `DEMO_SECRET__PASSWD=/secrets/secret.file` (without prefix) response
will contain the file name which was defined in the environment variable:

```bash

> # curl -v localhost:8080

Reviewer: If marking this as shell, don't include the prompt.

Author: Fixed

```

The output is similar to this:

```text
...
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.5.0
> Accept: */*
>
< HTTP/1.1 200 OK
...
/secrets/secret.file
```

```bash

> # strings /proc/1/environ

Reviewer: If marking this as shell, don't include the prompt.

Author: Fixed

```

```text
...
DEMO_SECRET__PASSWD=/secrets/secret.file
...
```

```bash

> # cat /secrets/secret.file

Reviewer: If marking this as shell, don't include the prompt.

```

```text
cat: /secrets/secret.file: No such file or directory
```

`/secrets/secret.file` is a string interpreted as sensitive data, that's why
the file wasn't injected.

The attacker's life has become more complicated because now in order to
access sensitive data it's necessary either to make a process dump or to use a
debugger tool. In order to show debug process, first it's necessary to make
build with disabled compiler optimization flags and run `gdb`:

```bash
go build -gcflags=all="-N -l" cmd/stage-03/main.go
echo 'MySecret' > secret.file

> DEMO_SECRET__PASSWD=file://secret.file gdb main

Reviewer: Better to [separate the command from sample output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

Author: Fixed

```

Let's check current process environment:

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

Author: Fixed

...
(gdb) show environment DEMO_SECRET__PASSWD
DEMO_SECRET__PASSWD = file://secret.file
...
```

Let's set a breakpoint and run debug. In other terminal tab or web browser it's
necessary to make request to `root`
location (the request is supposed to get stuck):

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

(gdb) break 48
Breakpoint 1 at 0x784cf5: file ..., line 48.
(gdb) run
```

Once the request is received, the process will be stopped (due to breakpoint).
`os.LookupEnv` function returns environment variable value if it is defined.

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

Author: Fixed

Thread 1 "main" hit Breakpoint 1, main.httpHandle (w=..., r=0xc0000c2000) at ...
48                      if envVal, isExistEnv = os.LookupEnv(envVar); isExistEnv {
```

Next,let's move on to checking for the file existence the path to which is
stored in the envVal variable:

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

Author: Fixed

(gdb) next
49                              if filePath, isExistFile = fileExist(envVal); isExistFile {
(gdb) print main.envVal
$1 = 0xc00001e0d4 "file://secret.file"
```

Next, `os.ReadFile` read file, `os.Setenv` makes override value of the
environment variable, but `os.Getenv` gets new value from recently-rewritten
environment variable:

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

Author: Fixed

(gdb) next
50                                      body, err := os.ReadFile(filePath)
...
54                                      os.Setenv(envVar, string(body))
(gdb) next
55                                      envVal = os.Getenv(envVar)
```

In conclusion, let's make sure, that the same variable has different values:

> ```bash

Reviewer: This isn't shell; please drop the `bash` suffix.

(gdb) next
56                                      changed = true
(gdb) print main.envVal
$2 = 0xc0000ea014 "MySecret\n"
(gdb) show environment DEMO_SECRET__PASSWD
DEMO_SECRET__PASSWD = file://secret.file
```

> As you can see debug process requires a few conditions such as: disabled

Reviewer: (nit)
```suggestion
As you can see, the debug process requires a few conditions such as: disabled
```
